src/daemon/db_util.rs

```diff
@@ -0,0 +1,190 @@
+use futures::TryStreamExt;
+use std::ffi::OsStr;
+use std::fs;
+use std::{
+    path::{Path, PathBuf},
+    time,
+};
+use tokio::io::AsyncWriteExt;
+use tracing::info;
+use url::Url;
+use walkdir::WalkDir;
+
+pub fn load_all_forest_cars<T>(
+    store: &ManyCar<T>,
+    forest_car_db_dir: impl AsRef<Path>,
+) -> anyhow::Result<()> {
+    let forest_car_db_dir = forest_car_db_dir.as_ref();
+    if !forest_car_db_dir.is_dir() {
+        fs::create_dir_all(forest_car_db_dir)?;
+    }
+    for file in WalkDir::new(forest_car_db_dir)
+        .into_iter()
+        .filter_map(|entry| {
+            if let Ok(entry) = entry {
+                if let Some(filename) = entry.file_name().to_str() {
+                    if filename.ends_with(FOREST_CAR_FILE_EXTENSION) {
+                        return Some(entry.into_path());
+                    }
+                }
+            }
+            None
+        })
+    {
+        match ForestCar::new(Mmap::map_path(&file)?) {
+            Ok(car) => {
+                store.read_only(car.into());
+                info!("Loaded car DB at {}", file.display());
+            }
+            Err(err) => {
+                bail!("Error loading car DB at {}: {err}", file.display())
+            }
+        };
+    }
+
+    Ok(())
+}
+
+/// This function validates and stores the CAR binary from `from_path`(either local path or URL) into the `{DB_ROOT}/car_db/`
+/// (automatically trans-code into `.forest.car.zst` format when needed), and returns its final file path and the heaviest tipset.
+pub async fn import_chain_as_forest_car(
+    from_path: &Path,
+    forest_car_db_dir: &Path,
+    consume_snapshot_file: bool,
+) -> anyhow::Result<(PathBuf, Tipset)> {
+    info!("Importing chain from snapshot at: {}", from_path.display());
+
+    let stopwatch = time::Instant::now();
+
+    let downloaded_car_temp_path =
+        tempfile::NamedTempFile::new_in(forest_car_db_dir)?.into_temp_path();
+    if let Ok(url) = Url::parse(&from_path.display().to_string()) {
+        download_to(url, &downloaded_car_temp_path).await?;
+    } else {
+        move_or_copy_file(from_path, &downloaded_car_temp_path, consume_snapshot_file)?;
+    }
+
+    let forest_car_db_path = forest_car_db_dir.join(format!(
+        "{}{FOREST_CAR_FILE_EXTENSION}",
+        chrono::Utc::now().timestamp_millis()
+    ));
+
+    if ForestCar::is_valid(&Mmap::map_path(&downloaded_car_temp_path)?) {
+        downloaded_car_temp_path.persist(&forest_car_db_path)?;
+    } else {
+        // Use another temp file to make sure all final `.forest.car.zst` files are complete and valid.
+        let forest_car_db_temp_path =
+            tempfile::NamedTempFile::new_in(forest_car_db_dir)?.into_temp_path();
+        transcode_into_forest_car(&downloaded_car_temp_path, &forest_car_db_temp_path).await?;
+        forest_car_db_temp_path.persist(&forest_car_db_path)?;
+    }
+
+    let ts = ForestCar::new(Mmap::map_path(&forest_car_db_path)?)?.heaviest_tipset()?;
```

Comment:
The use of mmap should be governed by the environment variable.

Reply:
Fixed